File: services/terrapod/api/routers/labels.py
"""Read-only labels browser API.

UX CONTRACT: consumed by `web/src/app/labels/page.tsx`. Changes to
response shapes here MUST be matched by frontend updates.

Endpoints:
    GET /api/v2/labels                       — distinct keys + per-type counts
    GET /api/v2/labels/{key}                 — distinct values for a key
    GET /api/v2/labels/{key}/{value}         — entities tagged with key=value

All three are RBAC-filtered: results only include labels carried by
entities the caller has at least `read` on for that entity's
permission model. Read-only by design — labels continue to be edited
on each entity's own edit page (no labels-admin surface here).
"""

from fastapi import APIRouter, Depends, Path
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from terrapod.api.dependencies import AuthenticatedUser, get_current_user
from terrapod.db.session import get_db
from terrapod.services import labels_service

router = APIRouter(prefix="/api/v2", tags=["labels"])


@router.get("/labels")
async def list_keys(
    user: AuthenticatedUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Return all label keys in use across readable entities.

    Each key entry includes the count of distinct values and a
    breakdown of how many entities of each type carry that key.
    """
    keys = await labels_service.aggregate_keys(db, user)
    return JSONResponse(content={"data": keys})


@router.get("/labels/{key}")
async def list_values(
    key: str = Path(..., min_length=1, max_length=200),
    user: AuthenticatedUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Return distinct values for `key`, with per-entity-type counts.

    Empty `data` is a valid response — means no readable entity
    carries this key (or no entity at all).
    """
    values = await labels_service.aggregate_values_for_key(db, user, key)
    return JSONResponse(content={"data": values})


@router.get("/labels/{key}/{value}")
async def list_entities(
    key: str = Path(..., min_length=1, max_length=200),
    value: str = Path(..., max_length=500),
    user: AuthenticatedUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> JSONResponse:
    """Return entities tagged with exactly `key=value`, grouped by type."""
    grouped = await labels_service.list_entities_for_label(db, user, key, value)
    return JSONResponse(content={"data": grouped})

File: go-terrapod/labels.go
package terrapod

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
)

// The labels browser is the read-only cross-entity label index served at
// /api/terrapod/v1/labels. All three endpoints are RBAC-filtered: results
// only include labels carried by entities the caller can read. Responses
// are plain JSON ({"data": ...}), not JSON:API resources.

// LabelKey is one distinct label key in use, with a count of distinct
// values and a per-entity-type breakdown of how many entities carry it.
type LabelKey struct {
	Key          string         `json:"key"`
	ValueCount   int            `json:"value-count"`
	EntityCounts map[string]int `json:"entity-counts"`
}

// LabelValue is one distinct value for a given key, with a per-entity-type
// count of how many entities carry key=value.
type LabelValue struct {
	Value        string         `json:"value"`
	EntityCounts map[string]int `json:"entity-counts"`
}

// LabelEntity is the minimal shape of an entity tagged with a label —
// enough to render a row and link to the entity's own page.
type LabelEntity struct {
	Type   string            `json:"type"`
	ID     string            `json:"id"`
	Name   string            `json:"name"`
	Labels map[string]string `json:"labels"`
}

// ListLabelKeys returns all label keys in use across readable entities.
func (c *Client) ListLabelKeys(ctx context.Context) ([]LabelKey, error) {
	data, err := c.Get(ctx, "/api/terrapod/v1/labels")
	if err != nil {
		return nil, err
	}
	var out struct {
		Data []LabelKey `json:"data"`
	}
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, fmt.Errorf("parse label keys: %w", err)
	}
	return out.Data, nil
}

// ListLabelValues returns the distinct values for a key. An empty slice is
// a valid response (no readable entity carries the key).
func (c *Client) ListLabelValues(ctx context.Context, key string) ([]LabelValue, error) {
	data, err := c.Get(ctx, "/api/terrapod/v1/labels/"+url.PathEscape(key))
	if err != nil {
		return nil, err
	}
	var out struct {
		Data []LabelValue `json:"data"`
	}
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, fmt.Errorf("parse label values: %w", err)
	}
	return out.Data, nil
}

// ListLabelEntities returns the entities tagged exactly key=value, grouped
// by entity type ("workspaces", "agent-pools", "registry-modules",
// "registry-providers"). Empty type lists are kept.
func (c *Client) ListLabelEntities(ctx context.Context, key, value string) (map[string][]LabelEntity, error) {
	path := fmt.Sprintf("/api/terrapod/v1/labels/%s/%s", url.PathEscape(key), url.PathEscape(value))
	data, err := c.Get(ctx, path)
	if err != nil {
		return nil, err
	}
	var out struct {
		Data map[string][]LabelEntity `json:"data"`
	}
	if err := json.Unmarshal(data, &out); err != nil {
		return nil, fmt.Errorf("parse label entities: %w", err)
	}
	return out.Data, nil
}
